tests/utils.py

```diff
@@ -1,54 +1,71 @@
+from __future__ import annotations
+
 import contextlib
-import http.server
-import pathlib
-import threading
+from http.server import ThreadingHTTPServer
+from pathlib import Path
 from ssl import PROTOCOL_TLS_SERVER, SSLContext
+from threading import Thread
+from typing import TYPE_CHECKING, TypeVar
 
 import filelock
 
+if TYPE_CHECKING:
+    from collections.abc import Callable, Generator
+    from contextlib import AbstractContextManager
+    from socketserver import BaseRequestHandler
+    from typing import Any, Final
+
 # Generated with:
 # $ openssl req -new -x509 -days 3650 -nodes -out cert.pem \
 #     -keyout cert.pem -addext "subjectAltName = DNS:localhost"
-TESTS_ROOT = pathlib.Path(__file__).parent
-CERT_FILE = str(TESTS_ROOT / "certs" / "cert.pem")
+TESTS_ROOT: Final[Path] = Path(__file__).parent
+CERT_FILE: Final[str] = str(TESTS_ROOT / "certs" / "cert.pem")
 
 # File lock for tests
-LOCK_PATH = str(TESTS_ROOT / 'test-server.lock')
+LOCK_PATH: Final[str] = str(TESTS_ROOT / 'test-server.lock')
 
 
-class HttpServerThread(threading.Thread):
-    def __init__(self, handler, *args, **kwargs):
+class HttpServerThread(Thread):
+    def __init__(self, handler: type[BaseRequestHandler], /, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
-        self.server = http.server.ThreadingHTTPServer(("localhost", 7777), handler)
+        self.server = ThreadingHTTPServer(("localhost", 7777), handler)
 
-    def run(self):
+    def run(self) -> None:
         self.server.serve_forever(poll_interval=0.001)
 
-    def terminate(self):
+    def terminate(self) -> None:
         self.server.shutdown()
         self.server.server_close()
         self.join()
 
 
 class HttpsServerThread(HttpServerThread):
-    def __init__(self, handler, *args, **kwargs):
+    def __init__(
+        self, handler: type[BaseRequestHandler], /, *args: Any, **kwargs: Any,
+    ) -> None:
         super().__init__(handler, *args, **kwargs)
         sslcontext = SSLContext(PROTOCOL_TLS_SERVER)
         sslcontext.load_cert_chain(CERT_FILE)
         self.server.socket = sslcontext.wrap_socket(self.server.socket, server_side=True)
 
 
-def create_server(thread_class):
-    def server(handler):
+_T_co = TypeVar('_T_co', bound=HttpServerThread, covariant=True)
+
+
+def create_server(
+    server_thread_class: type[_T_co],
+) -> Callable[[type[BaseRequestHandler]], AbstractContextManager[_T_co]]:
```

Comment:
I like the use of the `@contextlib.contextmanager` decorator, but I'm not so keen on the changes to these four lines.  I think it's less comprehensible to newcomers.

(yes, this method is kinda complicated, but we shouldn't make it look even more complicated)

Reply:
s/four/three/

Reply:
Mmmh. I'd say newcomers are not meant to play with this. But let's revert it (though, it's even worse imo to return the decorated function by calling explicitly the decorator).

Reply:
Wait, so you mean the typing? sorry but for that one, I want to make it a typed function. The reason is that I don't want it to be left untyped because otherwise `http_server` has `Any` type.

Reply:
Also, note that `create_server` is only used in this file for creating `http_server` and `https_server` but not in the rest of the test suite.

Reply:
Ok, go ahead.  I'll probably open a refactor PR to suggest something more readable after merge (but won't try to introduce changes in parallel with this PR).

Reply:
Not sure if it will be helpful in this specific case, but [`TypeAlias`](https://typing.readthedocs.io/en/latest/source/modernizing.html#type-aliases) can be useful for some cases like this (essentially a way to create a variable used for type hinting).  Python3.10+ / `typing_extensions` though.

Reply:
Yes, but I'm not sure I've already added the explicit `typing_extensions` dependency. Since we are anyway targetting Mypy for Python 3.9, it doesn't really matter I think.

Reply:
I tried various things - some of them changed the code a lot!

But ultimately the refactor I ended up with was [this](https://github.com/sphinx-doc/sphinx/commit/dc82b4d23e2de8d95fccef0e3f8d83e7e7b38300).

It doesn't change much, so -- unless any of it looks particularly useful -- I think we can skip it.  (full patch also included below)

<details>

<summary>click to display as patch</summary>

```patch
commit dc82b4d23e2de8d95fccef0e3f8d83e7e7b38300
Author: James Addison <jay@jp-hosting.net>
Date:   Sun Mar 17 12:42:15 2024 +0000

    [refactor] Attempt to make test HTTP server type hinting more comprehensible

diff --git a/tests/utils.py b/tests/utils.py
index 1fbd431eb..72e38fe62 100644
--- a/tests/utils.py
+++ b/tests/utils.py
@@ -5,16 +5,16 @@ from http.server import ThreadingHTTPServer
 from pathlib import Path
 from ssl import PROTOCOL_TLS_SERVER, SSLContext
 from threading import Thread
-from typing import TYPE_CHECKING, TypeVar
+from typing import TYPE_CHECKING
 
 import filelock
 
 if TYPE_CHECKING:
-    from collections.abc import Callable, Generator
-    from contextlib import AbstractContextManager
     from socketserver import BaseRequestHandler
     from typing import Any, Final
 
+    HandlerClass = type[BaseRequestHandler]  # TypeAlias
+
 # Generated with:
 # $ openssl req -new -x509 -days 3650 -nodes -out cert.pem \
 #     -keyout cert.pem -addext "subjectAltName = DNS:localhost"
@@ -26,7 +26,7 @@ LOCK_PATH: Final[str] = str(TESTS_ROOT / 'test-server.lock')
 
 
 class HttpServerThread(Thread):
-    def __init__(self, handler: type[BaseRequestHandler], /, *args: Any, **kwargs: Any) -> None:
+    def __init__(self, handler: HandlerClass, /, *args: Any, **kwargs: Any) -> None:
         super().__init__(*args, **kwargs)
         self.server = ThreadingHTTPServer(("localhost", 7777), handler)
 
@@ -40,23 +40,26 @@ class HttpServerThread(Thread):
 
 
 class HttpsServerThread(HttpServerThread):
-    def __init__(
-        self, handler: type[BaseRequestHandler], /, *args: Any, **kwargs: Any,
-    ) -> None:
+    def __init__(self, handler: HandlerClass, /, *args: Any, **kwargs: Any) -> None:
         super().__init__(handler, *args, **kwargs)
         sslcontext = SSLContext(PROTOCOL_TLS_SERVER)
         sslcontext.load_cert_chain(CERT_FILE)
         self.server.socket = sslcontext.wrap_socket(self.server.socket, server_side=True)
 
 
-_T_co = TypeVar('_T_co', bound=HttpServerThread, covariant=True)
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterator
+    from contextlib import AbstractContextManager as ContextManager
+    from typing import TypeVar
+
+    # Where HttpServerThread is accepted, HttpServerThread is also accepted; it is covariant.
+    # https://peps.python.org/pep-0484/#covariance-and-contravariance
+    Server = TypeVar("Server", bound=HttpServerThread, covariant=True)
 
 
-def create_server(
-    server_thread_class: type[_T_co],
-) -> Callable[[type[BaseRequestHandler]], AbstractContextManager[_T_co]]:
+def create_server(server_thread_class: type[Server]) -> Callable[[HandlerClass], ContextManager[Server]]:
     @contextlib.contextmanager
-    def server(handler_class: type[BaseRequestHandler]) -> Generator[_T_co, None, None]:
+    def server(handler_class: HandlerClass) -> Iterator[Server]:
         lock = filelock.FileLock(LOCK_PATH)
         with lock:
             server_thread = server_thread_class(handler_class, daemon=True)
```
</details>

Reply:
If you want to use the TypeAlias, it's either:

```
X = ...  # type: TypeAlias
```

or 

```
X: TypeAlias = ...
```

but I've never seen just `# TypeAlias`. By the way, I did not see that I could have put the `_T_co` in a TYPE_CHECKING block.